### PR TITLE
update test framework and tests to support 4x version of Pester

### PR DESCRIPTION
**Pull request by JamesWTruher**
Use the latest version of Pester from the Gallery to invoke our tests. A number of test changes needed to be made because some assertions were dropped in the 4x version of Pester.
I have left the Restore-PSPester function in the case that someone would like to have the older version. This can certainly be removed if desired.

### CODE REVIEW

**Comment by markekraus** [comment]
Maintainers: please make sure this passes `[feature]` tests before merging. 

@JamesWTruher can you please push a commit with `[feature]` in the commit message? Thanks!

**Comment by TravisEz13** [comment]
as @markekraus, can you please work with one of the maintainers to run your tests with feature in the CI system.

**Comment by stknohg** [comment]
Why don’t we adopt Pester 4.0 new assertion syntax?

e.g.

```powershell
# current syntax
$var1 | Should Not Be 10

# Pester 4.0 new assertion syntax
$var1 | Should -Not -Be 10
```

**Comment by markekraus** [comment]
@stknohg My preference would be to make the change to the new assertion syntax in separate PR's, one per test file, after we have successfully moved the tests to work with 4.0. It's one thing to make the tests work with 4.0, it is an entirely other thing to switch to the new syntax. switching all tests to the new syntax in this PR would make it hard to see what changes were necessary just to get it working with 4.0.

**Comment by stknohg** [comment]
@markekraus I think it's good. 👍

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-Variable.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-Variable.Tests.ps1]: be -> Be

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Set-PSBreakpoint.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Set-PSBreakpoint.Tests.ps1]: Above I see another pattern like `Should BeIn @(11,12,13)`. 
What pattern we'll use in tests?

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Isn't `BeIn` only used for validating if the left side is contained in the collection on the right side?  Don't understand why this isn't just `Be`?

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Isn't BeIn only used for validating if the left side is contained in the collection on the right side? Don't understand why this isn't just Be?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1]: Is this line indented correctly?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-Variable.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-Variable.Tests.ps1]: Wouldn't it be better to change this to:

```powershell
Remove-Variable $var1 -ErrorAction SilentlyContinue -ErrorVariable err
$err.FullyQualifiedErrorId | Should Be "VariableNotFOund,Microsoft.PowerShell.Commands.RemoveVariableCommand"
```

(applies to all the `Should throw` checks)

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

**Comment by bergmeister** [comment]
Can we please update one of the readme files with basic instructions and a link how to upgrade pester v3 to v4 (since this is not straightforward in most cases)?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-Variable.Tests.ps1
Comment (JamesWTruher) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-Variable.Tests.ps1]: I'll update the changes that I made to get it to work in this way, but I will defer changing all of these to a follow-on PR. My plan is to not update all tests with this PR

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Set-PSBreakpoint.Tests.ps1
Comment (JamesWTruher) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Set-PSBreakpoint.Tests.ps1]: I don't think it matters - both are valid expressions

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1
Comment (JamesWTruher) [reply test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: as it turns out, there are multiple loaded modules which are named TestModule, thus a comparison with "Be" fails because it's a collection. I didn't want to rearchitect the tests.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1
Comment (JamesWTruher) [reply test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: same as above

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1
Comment (JamesWTruher) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Variable.Tests.ps1]: there is a sad combination of tabs and spaces in this file, I'll fix up this line and save the whitespace fixing for a later PR

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Set-PSBreakpoint.Tests.ps1
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Set-PSBreakpoint.Tests.ps1]: We should probably just standardize on one of them for consistency/readability.  I prefer sans @(), myself.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Got it. Closed.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1]: Closed

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-Variable.Tests.ps1
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Remove-Variable.Tests.ps1]: Ok.  Closed.

**Review by SteveL-MSFT — APPROVED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Set-PSBreakpoint.Tests.ps1
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Set-PSBreakpoint.Tests.ps1]: I agree that we will request `@()` template for `Should BeIn` in future test reviews.

File: docs/testing-guidelines/WritingPesterTests.md
Comment (iSazonov) [inline docs/testing-guidelines/WritingPesterTests.md]: Speller test failed on "4x"

**Comment by iSazonov** [comment]
CI Apveyor temporary failed on Web tests. #6069 address the problem.

File: docs/testing-guidelines/WritingPesterTests.md
Comment (daxian-dbw) [reply docs/testing-guidelines/WritingPesterTests.md]: The corresponding error in Travis CI:
```
The command "pwsh -File tools/travis.ps1" exited with 0.
2.37s$ if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then mdspell '**/*.md' '!**/Pester/**/*.md' --ignore-numbers --ignore-acronyms --report; fi
    docs/testing-guidelines/WritingPesterTests.md
        5 |  Core is using Pester version 4x which has some changes from e 
>> 1 spelling error found in 67 files
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Invoke-Item.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Invoke-Item.Tests.ps1]: In line 67 we kill all notepad processes.

**Review by iSazonov — APPROVED** [review]
LGTM with one comment.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Invoke-Item.Tests.ps1
Comment (JamesWTruher) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Invoke-Item.Tests.ps1]: in CI this is probably ok - on a dev box, not so much - the fix for this should be part of the generalized test review

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**
